packages/playwright-core/src/server/browserContext.ts

```diff
@@ -610,11 +679,56 @@ export abstract class BrowserContext extends SdkObject {
         for (const originState of state.origins) {
           const frame = page.mainFrame();
           await frame.goto(metadata, originState.origin);
-          await frame.evaluateExpression(`
-            originState => {
-              for (const { name, value } of (originState.localStorage || []))
-                localStorage.setItem(name, value);
-            }`, { isFunction: true, world: 'utility' }, originState);
+
+          for (const dbInfo of (originState.indexedDB || [])) {
+            for (const store of dbInfo.stores) {
+              for (const record of store.records) {
+                if (record.key !== undefined)
+                  record.key = utilitySerializers.parseEvaluationResultValue(JSON.parse(record.key));
+                record.value = utilitySerializers.parseEvaluationResultValue(JSON.parse(record.value));
+              }
+            }
+          }
+
+          async function _restoreStorageState(originState: channels.OriginStorage) {
+            for (const { name, value } of (originState.localStorage || []))
+              localStorage.setItem(name, value);
+
+            function idbRequestToPromise<T extends IDBOpenDBRequest | IDBRequest>(request: T) {
```

Comment:
Same thing here, let's make a try/catch and provide an IndexedDB-related error message.